<h1><center>Laboratorio 2: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Diego Irarrázaval

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Joaquín Ortega Vásquez
- Nombre de alumno 2: María José Zambrano


### **Link de repositorio de GitHub:** `https://github.com/mzambrano1/Lab-programacion-cientifica`

### Indice 

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

# Temas a tratar

- Manejo de DataFrames con la libreria `pandas`.
- Visualizaciones interactivas de los datos con `plotly`.

## Reglas:

- Fecha de entrega: TBD
- **Grupos de 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer material y segmentos de código del curso que estimen conveniente.

### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


El laboratorio deberá ser desarrollado **sin el uso indiscriminado de nativos de python** (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

#Importamos librerias utiles 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.express as px

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán realizar un Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

In [2]:
# Usar solamente si utilizan Collab.
from google.colab import drive
drive.mount("/content/drive")
dir = '//content//drive//My Drive//lab2//'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
df_grades = pd.read_csv(dir+'students_grades.csv')


## 1.1 Revisión de datos [1 punto]

<p align="center">
  <img src="https://media2.giphy.com/media/4FQMuOKR6zQRO/giphy.gif?cid=ecf05e47ii9lhu9tc9tb31d6lafrysp2zyf6u8w3y348mk8t&rid=giphy.gif&ct=g">
</p>



El Rector *Don Caguayo* le comenta que actualmente la Universidad posee un total de **1000** alumnos inscritos. Sin embargo, al revisar su *Excel*, no le cabe en la cabeza que hayan más de **1000** alumnos en la universidad. 

Consternado, el rector cree que se han realizado malas prácticas al inscribir tanto los registros de las notas como las notas como tal y les solicita a ustedes (aka, los expertos) que revisen los datos. 

Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**.

Cree una función que le permita revisar si las diferentes variables que componen el dataset son coherentes y si es posible observar anomalías en estas.

> ¿Sera necesario limpiar los datos?, comente esto en la celda ubicada debajo de **justificación**. (agregar si están bien el rango de notas)

**To-Do**

Crear la función que `analisis_exploratorio` que:
- [ ] Imprima un extracto del Dataframe (usar `display` en vez de `print` para los DataFrames).
- [ ] Imprima las dimensiones del DataFrame.
- [ ] Describa los datos numéricos a través de ```.describe()```.
- [ ] Imprima el total de filas repetidas (hint: Buscar el método```duplicated()```.).
- [ ] Imprima la cantidad de valores nulos (`ìsna()`).
- [ ] Imprima la cantidad de valores unicos.

Además, genere visualizaciones (fuera de la función) que:
- [ ] Muestre la distribución de las variables numericas.

Por último, 

- [ ] Justifique si será necesario limpiar los datos y aplique los análisis univariado y/o bivariado vistos en clase.


In [4]:
def analisis_preliminar_datos(df_grades):
  print('Extracto del dataframe: ')
  display(df_grades)
  print('\n')
  print('dimensiones del dataframe: ',df_grades.shape,'\n')
  print('Datos númericos del dataframe: ')
  display(df_grades.describe())
  print('\n')

  print('filas duplicadas del dataframe: ')
  display(df_grades[df_grades.duplicated()])
  print('\n')
  print('cantidad de datos nulos por columna: ')
  print(df_grades.isna().sum(),'\n')
  print('valores unicos: ')
  print(df_grades.nunique())


In [5]:
analisis_preliminar_datos(df_grades)

Extracto del dataframe: 


names  gender  ... reading score writing score
0        Rita Courtney  female  ...          46.0          46.0
1      Madeline Fuller  female  ...         142.0         142.0
2        Daniel Rester    male  ...         196.0         193.0
3     Charles Linstrom    male  ...          78.0          81.0
4       Geraldine Hale  female  ...         166.0         166.0
...                ...     ...  ...           ...           ...
1395  Alexander Dillon    male  ...         142.0         193.0
1396    Marina Zeigler  female  ...          74.0          81.0
1397     Laurie Carter  female  ...          61.0          62.0
1398      Joseph Mccoy    male  ...         196.0         193.0
1399      Amanda Perez  female  ...          80.0          76.0

[1400 rows x 9 columns]



dimensiones del dataframe:  (1400, 9) 

Datos númericos del dataframe: 


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000



filas duplicadas del dataframe: 


names gender  ... reading score writing score
827     Justin Harris   male  ...          42.0          38.0
900   Anthony Malcomb   male  ...          28.0          27.0
1223  Charles Jackson   male  ...          26.0          22.0

[3 rows x 9 columns]



cantidad de datos nulos por columna: 
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64 

valores unicos: 
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64


> Utilice alguna visualización (gráfique) para observar el comportamiento de los datos numericos. ¿Estan todos los valores correctos?

In [6]:
# Group data together
df2 = df_grades.dropna()
import plotly.figure_factory as ff
hist_data = [df2['math score'], df2['reading score'] , df2['writing score']]

group_labels = ['math', 'reading', 'witring']

# Create distplot with custom bin_size
fig = ff.create_distplot(hist_data, group_labels, bin_size=.2)
fig.show()

**Justificación:**


Se evidencia que hay valores que no están buenos dado que fue necesario deshacerse de los valores nulos, mediante la función dropna. Con respecto la dsitribución se puede notar que los 3 puntajes tienen una distribuicón similar. 

 Dado que las notas solo pueden ir de 0 a 100, existen muchos datos incorrectos, porque se pueden ver puntajes muy superiores al máximo.

Es necesaria una  limpieza de los datos para eliiminar los nulos, datos repetidos y datos que no están correctos.


## 1.2 Limpieza de Datos [1 punto]

<p align="center">
  <img src="https://media3.giphy.com/media/c09yGbnSyaFRS/giphy.gif?cid=ecf05e47d0ejjjnpfj0ywcbwz7zk7sw2rfdpis890gmdpvj2&rid=giphy.gif&ct=g">
</p>


En base a lo realizado anteriormente, tome la decisión (si es necesario) de limpiar el dataset. Asuma los supuestos que estime conveniente para limpiar el DataFrame y cree una función que le permita limpiar los datos, si estima que no es necesario realizar esto, omita esta parte.

**To-Do:**

- [ ] Decidir si realizar o no la limpieza.
- [ ] Crear función que nos permita limpiar el dataset.
- [ ] Justifique las asunciones realizadas para limpiar los datos.

**Hint**

En el caso que decida limpiar el dataset, puede que le sean de utilidad las funciones ```.drop_duplicates()``` y ```.dropna()```.


In [7]:
df_grades = df_grades.dropna() 
df_grades = df_grades.drop_duplicates()
df_grades = df_grades[(df_grades['math score'] <= 100.0) & (df_grades['reading score'] <= 100.0) & (df_grades['writing score'] <= 100.0)]

In [8]:
analisis_preliminar_datos(df_grades)

Extracto del dataframe: 


names  gender  ... reading score writing score
0        Rita Courtney  female  ...          46.0          46.0
3     Charles Linstrom    male  ...          78.0          81.0
5          Brian Young    male  ...          66.0          63.0
6       Howard Jimenez    male  ...          75.0          76.0
7         Wayne Wilson    male  ...          86.0          87.0
...                ...     ...  ...           ...           ...
1392     Richard Young    male  ...          75.0          71.0
1394     Wanda Russell  female  ...          44.0          36.0
1396    Marina Zeigler  female  ...          74.0          81.0
1397     Laurie Carter  female  ...          61.0          62.0
1399      Amanda Perez  female  ...          80.0          76.0

[1000 rows x 9 columns]



dimensiones del dataframe:  (1000, 9) 

Datos númericos del dataframe: 


math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000



filas duplicadas del dataframe: 


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []



cantidad de datos nulos por columna: 
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64 

valores unicos: 
names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64


In [9]:
df2 = df_grades.dropna()
import plotly.figure_factory as ff
hist_data = [df2['math score'], df2['reading score'] , df2['writing score']]

group_labels = ['math', 'reading', 'witring']

# Create distplot with custom bin_size
fig = ff.create_distplot(hist_data, group_labels, bin_size=.2)
fig.show()

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

**Justificación:**

Luego de hacer la limpieza se obtiene un total de 1000 alumnos que corresponde con lo que esperaba el director. Se puede apreciar de mejor manera la distribución de los datos debido a que no están los datos erróneos.

## 1.3 Análisis exploratorio de datos



<p align="center">
  <img src="https://media4.giphy.com/media/ko7twHhomhk8E/giphy.gif?cid=ecf05e47xmgfz8ftaom8saa4cwktxpb0fkui9yk3nakb4ff1&rid=giphy.gif&ct=g">
</p>


### 1.3.1 Transformación de notas a "Nota Chilena" [0.5 punto]

En la ultima reunion con Don Caguayo este les comenta es complejo para el y su equipo entender la graduación de notas de 0 a 100. Por ende, le comenta que desde hace harto tiempo tienen la intención de *trnasformar* todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree un función que le permita pasar a la escala Chilena las notas que se escapan del rango [1,7]. 

(No ocupe ningún tipo de ciclo).

**To-Do**

- [ ] Tranforme los valores a la escala chilena.
- [ ] Asigne los valores transformados al DataFrame.

In [10]:
# en el gráfico de más arriba se ve que ninguna nota escapa del rango [1,7]
df_grades['math score'] =  df_grades['math score']*0.06 + 1
df_grades['reading score'] =  df_grades['reading score']*0.06 + 1
df_grades['writing score'] =  df_grades['writing score']*0.06 + 1

In [11]:
display(df_grades)

names  gender  ... reading score writing score
0        Rita Courtney  female  ...          3.76          3.76
3     Charles Linstrom    male  ...          5.68          5.86
5          Brian Young    male  ...          4.96          4.78
6       Howard Jimenez    male  ...          5.50          5.56
7         Wayne Wilson    male  ...          6.16          6.22
...                ...     ...  ...           ...           ...
1392     Richard Young    male  ...          5.50          5.26
1394     Wanda Russell  female  ...          3.64          3.16
1396    Marina Zeigler  female  ...          5.44          5.86
1397     Laurie Carter  female  ...          4.66          4.72
1399      Amanda Perez  female  ...          5.80          5.56

[1000 rows x 9 columns]

### 1.3.2 Obtención de promedio de alumnos [0.5 puntos]

Con todas las notas adjuntadas, la prestigiosa *Universidad De la Cachaña* necesita obtener el promedio general de cada uno de los alumnos (Grade Point Average - GPA). Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del DataFrame original.

**To-Do:**
- [ ] Obtenga el promedio de cada uno de los alumnos.
- [ ] Anexe esta información en el DataFrame original.


In [12]:
df_grades['promedio'] = (df_grades['writing score'] + df_grades['reading score'] + df_grades['math score'])/3

In [13]:
display(df_grades)

names  gender  ... writing score promedio
0        Rita Courtney  female  ...          3.76     3.58
3     Charles Linstrom    male  ...          5.86     5.78
5          Brian Young    male  ...          4.78     5.04
6       Howard Jimenez    male  ...          5.56     5.64
7         Wayne Wilson    male  ...          6.22     6.34
...                ...     ...  ...           ...      ...
1392     Richard Young    male  ...          5.26     5.30
1394     Wanda Russell  female  ...          3.16     3.06
1396    Marina Zeigler  female  ...          5.86     5.42
1397     Laurie Carter  female  ...          4.72     4.54
1399      Amanda Perez  female  ...          5.56     5.48

[1000 rows x 10 columns]

### 1.3.3 Reprobados [1 punto]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena.

Cree una función que permita obtener dos Dataframes, el primero debe poseer solamente los alumnos que han reprobado en la prestigiosa universidad y el segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.

**To-Do:**

- [ ] Obtenga una máscara DataFrame con los alumnos que reprobaron el año.
- [ ] Reemplaze todos los GPA reprobados con una **R**.

In [14]:
def reprobados(df_grades):
  df_reprobados = df_grades[df_grades['promedio'] < 4.0]
  df_aprobados = df_grades.copy()
  promedios = df_aprobados['promedio']
  promedios = promedios.where(promedios >= 4.0, 'R')
  df_aprobados['promedio'] = promedios
  return df_aprobados, df_reprobados

In [15]:
aprobados, reprobados = reprobados(df_grades)
display(aprobados)

names  gender  ... writing score promedio
0        Rita Courtney  female  ...          3.76        R
3     Charles Linstrom    male  ...          5.86     5.78
5          Brian Young    male  ...          4.78     5.04
6       Howard Jimenez    male  ...          5.56     5.64
7         Wayne Wilson    male  ...          6.22     6.34
...                ...     ...  ...           ...      ...
1392     Richard Young    male  ...          5.26      5.3
1394     Wanda Russell  female  ...          3.16        R
1396    Marina Zeigler  female  ...          5.86     5.42
1397     Laurie Carter  female  ...          4.72     4.54
1399      Amanda Perez  female  ...          5.56     5.48

[1000 rows x 10 columns]

In [16]:
display(reprobados)

names  gender  ... writing score promedio
0       Rita Courtney  female  ...          3.76     3.58
13        Mikki Amaya  female  ...          3.46     3.46
26      Samuel Dougal    male  ...          3.76     3.68
55    Rodger Talavera    male  ...          2.62     2.84
63     Morris Johnson    male  ...          4.18     3.96
...               ...     ...  ...           ...      ...
1371   Charles Harian    male  ...          3.94     3.92
1373    Wendy Gunnell  female  ...          2.92     2.86
1389      Diana Costa  female  ...          3.58     3.60
1391      Amy Sieving  female  ...          2.80     2.76
1394    Wanda Russell  female  ...          3.16     3.06

[104 rows x 10 columns]

### 1.3.4 Información por Categorías [1 punto]

Motivado por saber más sobre los datos, el Rector les solicita que genere una función que le permita comparar los promedios de diferentes **atributos categóricos** del DataFrame. Para esto su función deberá tener como entrada el DataFrame con todos los promedios (sin agregar las R's) y calcular el promedio **agrupando por etnias y el nivel educacional de los padres**.

**To-Do:**

- [ ] Hacer una función que nos permita obtener la comparación de promedios para dos atributos categóricos.
- [ ] Cada uno de los grupos expuestos deben exponer el porcentaje de alumnos que representan en el colegio.
- [ ] Comente lo que observa de los resultados.
- [ ] Visualice los datos generados con [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ).


In [17]:
def agrupar_etnias_padres(df_grades):
  n = df_grades.shape[0]
  grouped = df_grades.groupby(['race/ethnicity','parental level of education'])
  promedio = grouped.mean()
  promedio['porcentaje'] = grouped.count()['names'] * 100 / n
  return promedio

In [18]:
display(agrupar_etnias_padres(df_grades))

math score  ...  porcentaje
race/ethnicity parental level of education              ...            
group A        associate's degree             4.660000  ...         1.4
               bachelor's degree              5.030000  ...         1.2
               high school                    4.626667  ...         1.8
               master's degree                4.460000  ...         0.3
               some college                   4.833333  ...         1.8
               some high school               4.535000  ...         2.4
group B        associate's degree             4.965854  ...         4.1
               bachelor's degree              5.158000  ...         2.0
               high school                    4.587500  ...         4.8
               master's degree                5.030000  ...         0.6
               some college                   4.791351  ...         3.7
               some high school               4.708947  ...         3.8
group C        associate's degree             5.003846  ...         7.8
               bachelor's degree              5.089000  ...         4.0
               high school                    4.654375  ...         6.4
               master's degree                5.023158  ...         1.9
               some college                   4.907826  ...         6.9
               some high school               4.633061  ...         4.9
group D        associate's degree             5.056000  ...         5.0
               bachelor's degree              5.054286  ...         2.8
               high school                    4.771818  ...         4.4
               master's degree                5.351304  ...         2.3
               some college                   5.123881  ...         6.7
               some high school               5.005600  ...         5.0
group E        associate's degree             5.493846  ...         3.9
               bachelor's degree              5.593333  ...         1.8
               high school                    5.246364  ...         2.2
               master's degree                5.477500  ...         0.8
               some college                   5.429714  ...         3.5
               some high school               5.326667  ...         1.8

[30 rows x 5 columns]

**Repsuesta Esperada:**

| race/ethnicity | parental level of education | math score | reading score | writing score |  GPA | percentaje % |
|---------------:|----------------------------:|-----------:|--------------:|--------------:|-----:|-------------:|
|     group A    |      associate's degree     |       4.66 |          5.02 |          4.81 | 4.83 |          1.4 |
|                |      bachelor's degree      |       5.03 |          5.08 |          5.10 | 5.07 |          1.2 |
|                |         high school         |       4.63 |          4.77 |          4.63 | 4.68 |          1.8 |
|                |       master's degree       |       4.46 |          4.88 |          5.06 | 4.80 |          0.3 |
|                |         some college        |       4.83 |          4.95 |          4.90 | 4.89 |          1.8 |
|                |       some high school      |       4.54 |          4.72 |          4.51 | 4.59 |          2.4 |
|     group B    |      associate's degree     |       4.97 |          5.18 |          5.09 | 5.08 |          4.1 |
|                |      bachelor's degree      |       5.16 |          5.38 |          5.30 | 5.28 |          2.0 |
|                |         high school         |       4.59 |          4.81 |          4.68 | 4.69 |          4.8 |
|                |       master's degree       |       5.03 |          5.81 |          5.63 | 5.49 |          0.6 |
|                |         some college        |       4.79 |          4.95 |          4.85 | 4.86 |          3.7 |
|                |       some high school      |       4.71 |          4.99 |          4.88 | 4.86 |          3.8 |
|     group C    |      associate's degree     |       5.00 |          5.27 |          5.22 | 5.16 |          7.8 |
|                |      bachelor's degree      |       5.09 |          5.54 |          5.55 | 5.39 |          4.0 |
|                |         high school         |       4.65 |          4.87 |          4.70 | 4.74 |          6.4 |
|                |       master's degree       |       5.02 |          5.23 |          5.17 | 5.14 |          1.9 |
|                |         some college        |       4.91 |          5.17 |          5.13 | 5.07 |          6.9 |
|                |       some high school      |       4.63 |          4.94 |          4.80 | 4.79 |          4.9 |
|     group D    |      associate's degree     |       5.06 |          5.23 |          5.19 | 5.16 |          5.0 |
|                |      bachelor's degree      |       5.05 |          5.21 |          5.31 | 5.19 |          2.8 |
|                |         high school         |       4.77 |          4.86 |          4.79 | 4.81 |          4.4 |
|                |       master's degree       |       5.35 |          5.63 |          5.78 | 5.59 |          2.3 |
|                |         some college        |       5.12 |          5.25 |          5.30 | 5.23 |          6.7 |
|                |       some high school      |       5.01 |          5.20 |          5.15 | 5.12 |          5.0 |
|     group E    |      associate's degree     |       5.49 |          5.43 |          5.39 | 5.44 |          3.9 |
|                |      bachelor's degree      |       5.59 |          5.49 |          5.52 | 5.54 |          1.8 |
|                |         high school         |       5.25 |          5.22 |          5.05 | 5.17 |          2.2 |
|                |       master's degree       |       5.48 |          5.93 |          5.83 | 5.74 |          0.8 |
|                |         some college        |       5.43 |          5.36 |          5.21 | 5.33 |          3.5 |
|                |       some high school      |       5.33 |          5.17 |          4.99 | 5.16 |          1.8 |

> Visualice utilizando [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ) el DataFrame obtenido en esta sección. 

In [59]:
df_agrupado = agrupar_etnias_padres(df_grades).reset_index()
df_agrupado['promedio'] = df_agrupado['promedio'].round(decimals = 2)
fig = px.sunburst(df_agrupado, path = ['parental level of education','promedio','race/ethnicity'], values = 'porcentaje')
fig.show()

In [61]:
fig = px.sunburst(df_agrupado, path = ['race/ethnicity','parental level of education','promedio'], values = 'promedio')
fig.show()

> Comente los resultados obtenidos. ¿Hay alguna diferencia entre los grupos graficados?, ¿Es útil esta visualización?

**Justifique:**

De las visualizaciones anteriores se puede observar que los hijos de padres con grado de magister corresponden a un pequeño porcentaje del total. Además estos suelen tener mejores notas que el resto de los alumnos.

En cuanto a los grupos A,B,C,D y E, el grupo A es el que tiene peores notas, mientras que el E tiene mejores.

### 1.3.5 Combinar Dataset [1 punto]

Al notificarle los resultados obtenidos en la etapa de descubrimiento a Don Caguayo, este les informa que ha encontrado un Excel con nueva información, en esta se tienen los valores de dos asignaturas y desea que la unan con el dataset original (```students_grades.csv```).

Busque la forma de unir ambos DataFrames, para esto deberá anexar las columnas de ```history score``` y ```science score``` al final del DataFrame original.

**To-Do**

- [ ] Unir DataFrame ```students_grades.csv``` y ```other_grades.csv```.

In [62]:
df_other_grades = pd.read_csv(dir+'other_grades.csv')

In [64]:
df_grades.merge(df_other_grades, right_on = 'names', left_on = 'names')

names  gender  ... science score history score
0       Rita Courtney  female  ...            25           3.8
1    Charles Linstrom    male  ...            56           5.2
2         Brian Young    male  ...            58           3.8
3      Howard Jimenez    male  ...            72           5.0
4        Wayne Wilson    male  ...            94           4.6
..                ...     ...  ...           ...           ...
995     Richard Young    male  ...            62           4.2
996     Wanda Russell  female  ...            16           3.6
997    Marina Zeigler  female  ...            59           4.1
998     Laurie Carter  female  ...            43           4.3
999      Amanda Perez  female  ...            61           5.3

[1000 rows x 12 columns]

> Explique (pensando en una operación de conjuntos) que hace la operación de combinación por defecto (`inner`).

> Explique que hubiese sucedido si aplicaramos los siguientes merge en las variables anteriores:

- `left`
- `right`
- `outer`

**Justificación:**

- La operación por defecto inner, se combinan los valores de la intersección entre dos conjuntos, en este caso los alumnos que estaban en ambos archivos de notas.

- En caso de haber utilizado left, se combinan los valores de la intersección más los del conjunto de la izquierda que no estaban en la intersección, quedando estos últimos con valores none.

- Rigth es lo mismo que left, solo que conservando los valores de la derecha.

- Outer excluye la intersección entre los dos conjuntos, dejando solo los valores que no se repiten, quedando con none los valores faltantes.

## 1.4 Más visualizaciones [Bonus] [1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




### 1.4.1 Rendimiento por Asignatura

Con el DataFrame de rendimiento para cada uno de los alumnos grafique como distribuyen los datos para cada uno de los puntajes obtenidos en cada una de las asignaturas.

Para realizar esta tarea se les recomienda que utilicen el comando ```px.histogram``` de plotly. Para esto puede ser útil procesar los datos con un melt, agrupando en una columna las asignaturas y en otra columna la nota obtenida en dicha asignatura.

Ejemplo del melt:

|    Variable   | Value |
|:-------------:|:-----:|
| math score    | 6.4   |
| math score    | 5.2   |
| ...           | ...   |
| writing score | 4.5   |

**To-Do:**

- [ ] Utilice `plotly` para visualizar el rendimiento para los diferentes grupos que componen el dataset.
- [ ] Responder a la pregunta, ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?


In [85]:
df_melt = pd.melt(df_grades, id_vars = ['names'], value_vars = ['math score', 'reading score','writing score'])
display(df_melt)

names       variable  value
0        Rita Courtney     math score   3.22
1     Charles Linstrom     math score   5.80
2          Brian Young     math score   5.38
3       Howard Jimenez     math score   5.86
4         Wayne Wilson     math score   6.64
...                ...            ...    ...
2995     Richard Young  writing score   5.26
2996     Wanda Russell  writing score   3.16
2997    Marina Zeigler  writing score   5.86
2998     Laurie Carter  writing score   4.72
2999      Amanda Perez  writing score   5.56

[3000 rows x 3 columns]

In [91]:
fig = px.histogram(df_melt, x="value", color="variable", barmode = 'overlay')
fig.show()

In [86]:
fig = px.histogram(df_melt, x="value", color="variable")
fig.show()

> ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?

**Comente:**

La asignatura que más le cuesta a los alumnos es matemáticas, esto se puede ver al estudiar los hitogramas, notando que la distribución de 'math' está orrida un poco más a la izquierda que las demás.

### 1.4.2 Otras Visualizaciones

Genere dos visualizaciones extras que encuentre interesantes con estos datos. Agrupe los atributos que estime convenientes.


**To-Do:**
- [ ] Generar dos nuevas visualizaciones con los datos.

In [94]:
df_melt2 = pd.melt(df_grades, id_vars = ['names','gender'], value_vars = ['promedio'])
display(df_melt2)

names  gender  variable  value
0       Rita Courtney  female  promedio   3.58
1    Charles Linstrom    male  promedio   5.78
2         Brian Young    male  promedio   5.04
3      Howard Jimenez    male  promedio   5.64
4        Wayne Wilson    male  promedio   6.34
..                ...     ...       ...    ...
995     Richard Young    male  promedio   5.30
996     Wanda Russell  female  promedio   3.06
997    Marina Zeigler  female  promedio   5.42
998     Laurie Carter  female  promedio   4.54
999      Amanda Perez  female  promedio   5.48

[1000 rows x 4 columns]

In [97]:
fig = px.histogram(df_melt2, x="value", color="gender", barmode = 'overlay')
fig.show()

En la visualización anterior se muestra la distribución de los promedios obtenidos por hombres y mujeres. En ella se puede observar, a priori, que a las mujeres les fue un poco mejor que a los hombres, debido a que el histograma está más corrido hacia la derecha.

In [107]:
fig = px.box(df_grades, x="parental level of education", y="promedio", points="all")
fig.show()

En la visualización anterior se muestran gráficos de caja del promedio obtenido por los estudiantes, según el nivel educacional de sus padres. En ella se pude ver que los hijos de personas que no terminaron la escuela secundaria tienen mayor número de ocurrencias en notas bajas y una varianza mayor al resto de las categorías. Por el contrario los hijos de personas con grado de magister tienen mejores notas. Lo anterior no es concluyente, debido a que el número de personas de cada grupo está desbalanceado y por lo tanto no son comparables.

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>